lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/AffixCondition.java

```diff
@@ -0,0 +1,214 @@
+      }
+
+      int stripRangeStart = kind == PREFIX ? 0 : charPatterns.size() - strip.length();
+      int stripRangeEnd = kind == PREFIX ? strip.length() : charPatterns.size();
+      if (!strip.isEmpty()
+          && !strip.matches(unitePatterns(charPatterns.subList(stripRangeStart, stripRangeEnd)))) {
+        return null;
+      }
+
+      if (kind == PREFIX) {
+        return unitePatterns(charPatterns.subList(stripRangeEnd, charPatterns.size()));
+      }
+      return unitePatterns(charPatterns.subList(0, stripRangeStart));
+    } catch (PatternSyntaxException e) {
+      throw new IllegalArgumentException("On line " + line, e);
+    }
+  }
+
+  /** Produces a regexp from groups returned by {@link #parse} */
+  private static String unitePatterns(List<String> charPatterns) {
+    return charPatterns.stream()
+        .map(s -> s.length() == 1 && "()?$^{}*+|\\".indexOf(s.charAt(0)) >= 0 ? "\\" + s : s)
+        .collect(Collectors.joining());
+  }
+
+  /** Splits condition into small regexps, each standing for a single char to be matched. */
+  private static List<String> parse(String condition) {
+    List<String> groups = new ArrayList<>();
+    for (int i = 0; i < condition.length(); i++) {
+      char c = condition.charAt(i);
+      if (c == '[') {
+        int closing = condition.indexOf(']', i + 1);
+        if (closing <= 0) {
+          groups.add(condition.substring(i) + "]");
+          break;
+        }
+
+        groups.add(condition.substring(i, closing + 1));
+        i = closing;
+      } else if (c == '.') {
+        groups.add(".");
+      } else {
+        groups.add(String.valueOf(c));
+      }
+    }
+    return groups;
+  }
+
+  private static AffixCondition substringCondition(AffixKind kind, String stemCondition) {
+    boolean forSuffix = kind == AffixKind.SUFFIX;
+    int condLength = stemCondition.length();
+    return (word, offset, length) -> {
+      if (length < condLength) {
+        return false;
+      }
+      int matchStart = forSuffix ? offset + length - condLength : offset;
+      for (int i = 0; i < condLength; i++) {
+        if (stemCondition.charAt(i) != word[matchStart + i]) {
+          return false;
+        }
+      }
+      return true;
+    };
+  }
+
+  private static AffixCondition regexpCondition(AffixKind kind, String stemCondition) {
+    boolean forSuffix = kind == AffixKind.SUFFIX;
+    int condLength = parse(stemCondition).size();
+    CharacterRunAutomaton automaton =
+        new CharacterRunAutomaton(conditionRegexp(stemCondition).toAutomaton());
+    return (word, offset, length) ->
+        length >= condLength
+            && automaton.run(word, forSuffix ? offset + length - condLength : offset, condLength);
+  }
+
+  private static RegExp conditionRegexp(String regex) {
+    try {
+      return new RegExp(regex, RegExp.NONE);
+    } catch (IllegalArgumentException e) {
+      if (e.getMessage().contains("expected ']'")) {
+        return conditionRegexp(regex + "]");
```

Comment:
what is this? :)

Reply:
A hack (moved from Dictionary) to mimic Hunspell's behavior, which (probably accidentally) allows [ ] groups to be unclosed

Reply:
This comment made me think, and it seems this whole class can be rewritten somewhat shorter and more straightforward. I'll try that tomorrow.

Reply:
Maybe my other comment on the "unite" method is related. Ideally we could avoid RegExp class. RegExp class just parses and then uses automaton API to build the result. But if we bypass the parsing and build the DFA we want explicitly, we may be able to avoid some hassles.

Reply:
RegExp seems convenient, a one-liner to get the necessary result. I didn't intend to get rid of it. But if building the automaton for a regexp subset with just `.`, `[chars]` and `[^chars]` is super-easy, I could try that.